win32ss/gdi/eng/stretchblt.c

```diff
@@ -397,10 +534,169 @@ IntEngStretchBlt(SURFOBJ *psoDest,
+        if ((cyDest >= 0) && (cySrc < 0))
+        {
+            cyDest = -cyDest;
+            cySrc = -cySrc;
+        }
+
+        /* if the cyDest < 0 and cySrc >= 0 then we reverse */
+        /* the top and bottom values                        */
+        else if ((cyDest <= 0) && (cySrc > 0))
+        {
+            lTmp = SourceRect->top;
+            SourceRect->top = SourceRect->bottom;
+            SourceRect->bottom = lTmp;
+        }
+    }
+
+    /*****************************************************************************************************************************
+      We want to send the flip state to IntEngBitBlt without changing its parameter list.
+      So a way that we can do this is to use the DestRect to carry this information along with it.
+      Since there are four values, we can use their relative positions (coordinates) to indicate the four flip conditions.
+      if the cx = 0 then there can be no x-flip and if cy = 0 there can be no y-flip. We have no concerns about these conditions.
+      So we can set the four flip conditions as follows:
+
+      flip = 0 means no flips                             therefore left < right and top < bottom (normal well-formed rect)
+      flip = 1 means that we are flipped about the y-axis therefore left < right and top > bottom
+      flip = 2 means that we are flipped about the x-axis therefore left > right and top < bottom
+      flip = 3 means that we are flipped about both axes  therefore left > right and top > bottom
+    *****************************************************************************************************************************/
+
+    /* Make Well Ordered to start */
+    OutputRect = *DestRect;
+    RECTL_vMakeWellOrdered(&OutputRect);
+    *DestRect = OutputRect;
+
+    /* Here we do the tests and set our flip based on input conditions */
+    if (((cxSrc < 0) && (cxDest < 0)) || ((cxSrc >= 0) && (cxDest >= 0)))
+    {
+        flipy = FALSE;
+    }
+    else
+    {
+        flipy = TRUE;
+    }
+
+    if (((cySrc < 0) && (cyDest < 0)) || ((cySrc >= 0) && (cyDest >= 0)))
+    {
+        flipx = FALSE;
+    }
+    else
+    {
+        flipx = TRUE;
+    }
+
+    DPRINT("flip about x-axis is '%d' and flip and y-axis is '%d'.\n", flipx, flipy);
+
+    if (!flipx && !flipy)
+    {
+        flip = 0;
+    }
+    else if (!flipx && flipy)
+    {
+        flip = 1;
+    }
+    else if (flipx && !flipy)
+    {
+        flip = 2;
+    }
+    else
+    {
+        flip = 3;
+    }
+
+    DPRINT("flip is '%d'.\n", flip);
+
     /* Check if source and dest size are equal */
-    if (((DestRect->right - DestRect->left) == (SourceRect->right - SourceRect->left)) &&
-        ((DestRect->bottom - DestRect->top) == (SourceRect->bottom - SourceRect->top)))
+    if ((abs(DestRect->right - DestRect->left) == abs(SourceRect->right - SourceRect->left)) &&
+        (abs(DestRect->bottom - DestRect->top) == abs(SourceRect->bottom - SourceRect->top)))
```

Comment:
The addition of the "abs" functions should not be required. The DestRect was made well-ordered about 30 lines above.